In [3]:
import pandas as pd
import numpy as np

## Question 1

 - What's the version of NumPy that you installed?

 - You can get the version information using the __version__ field:

In [4]:
np.__version__

'1.20.1'

### Reading the dataset

In [5]:
data = pd.read_csv('data.csv')

In [6]:
data.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

## Question 2

 - How many records are in the dataset?

In [7]:
data.shape

(11914, 16)

We have `11914` records in this data.
 - Ans is `C`

## Question 3

 - Who are the most frequent car manufacturers (top-3) according to the dataset?

In [17]:
data.groupby('Make')['Model'].count().reset_index().sort_values(by='Model', ascending=False).head(3)

Make  Model
9    Chevrolet   1123
14        Ford    881
46  Volkswagen    809

 - Ans is `D` Chevrolet, Ford, Volkswagen.

## Question 4

 - What's the number of unique Audi car models in the dataset?

In [22]:
data[data.Make=='Audi']['Model'].nunique()

34

- Ans is `D` 34.

## Question 5

How many columns in the dataset have missing values?

In [27]:
data.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

- Ans is `A` 5.

## Question 6

1. Find the median value of "Engine Cylinders" column in the dataset.
2. Next, calculate the most frequent value of the same "Engine Cylinders".
3. Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
4. Now, calculate the median value of "Engine Cylinders" once again.

Has it changed?

In [31]:
data['Engine Cylinders'].median()

6.0

In [33]:
data['Engine Cylinders'].mode()[0]

4.0

In [35]:
data['Engine Cylinders'].fillna(data['Engine Cylinders'].mode()[0]).median()

6.0

- Ans is `B` No.

## Question 7

1. Select all the "Lotus" cars from the dataset.
2. Select only columns "Engine HP", "Engine Cylinders".
3. Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

What's the value of the first element of w?

In [41]:
lotus_cars = data[data.Make=='Lotus']

In [45]:
lotus_cars_unique = lotus_cars[['Engine HP', 'Engine Cylinders']].drop_duplicates()

In [49]:
x = lotus_cars_unique.to_numpy()

In [50]:
np.transpose(x)

array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [51]:
xt = np.transpose(x)

In [52]:
np.matmul(xt, x)

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [53]:
xtx = np.matmul(xt, x)

In [54]:
np.linalg.inv(xtx)

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [55]:
inverse = np.linalg.inv(xtx)

In [56]:
y = [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]

In [57]:
transpose_xtx = np.transpose(xtx)

In [58]:
np.matmul(inverse, xt)

array([[-1.31950096e-03,  2.84443321e-04,  2.29134897e-04,
        -4.18763778e-03,  4.46417989e-03, -2.39406462e-03,
         1.42221660e-03,  2.44147184e-03,  1.50122864e-03],
       [ 8.60893170e-02,  7.36567735e-04,  3.67976598e-03,
         2.54588185e-01, -2.13748621e-01,  1.51207962e-01,
        -5.18727169e-02, -1.14048164e-01, -6.40137937e-02]])

In [59]:
mul = np.matmul(inverse, xt)

In [60]:
w = np.matmul(mul, y)

In [61]:
w

array([  4.59494481, -63.56432501])

- Ans is `B` 4.5949.